src/presetsController.ts

```diff
@@ -0,0 +1,844 @@
+    } catch (e) {
+      log.error(localize('failed.to.parse', 'Failed to parse {0}: {1}', path.basename(file), util.errorToString(e)));
+      return undefined;
+    }
+    return presetsFile;
+  }
+
+  private async validatePresetsFile(presetsFile: preset.PresetsFile | undefined, file: string) {
+    if (!presetsFile) {
+      return undefined;
+    }
+    if (presetsFile.version < 2) {
+      await this.showPresetsFileVersionError(file);
+      return undefined;
+    }
+    const validator = await loadSchema('schemas/CMakePresets-schema.json');
+    const is_valid = validator(presetsFile);
+    if (!is_valid) {
+      const errors = validator.errors!;
+      log.error(localize('invalid.file.error', 'Invalid kit contents {0} ({1}):', path.basename(file), file));
+      for (const err of errors) {
+        log.error(` >> ${err.dataPath}: ${err.message}`);
+      }
+      return undefined;
+    }
+    log.info(localize('successfully.validated.presets', 'Successfully validated presets kits in {0}', file));
+    return presetsFile;
+  }
+
+  private async showPresetsFileVersionError(file: string): Promise<void> {
+    const useKitsVars = localize('use.kits.variants', 'Use kits and variants');
+    const changePresets = localize('edit.presets', 'Locate');
+    const result = await vscode.window.showErrorMessage(
+        localize('presets.version.error', 'CMakePresets version 1 is not supported. How would you like to proceed?'),
+        useKitsVars, changePresets);
+    if (result === useKitsVars) {
+      vscode.workspace.getConfiguration('cmake', this.folder.uri).update('useCMakePresets', false);
+    } else {
+      vscode.workspace.openTextDocument(vscode.Uri.file(file));
+    }
+  }
+
+  // Note: in case anyone want to change this, presetType must match the corresponding key in presets.json files
+  async addPresetAddUpdate(newPreset: preset.ConfigurePreset, presetType: 'configurePresets' | 'buildPresets' | 'testPresets') {
+    const originalPresetsFile: preset.PresetsFile = preset.getOriginalPresetsFile() || { version: 2 };
+    if (!originalPresetsFile[presetType]) {
+      originalPresetsFile[presetType] = [];
+    }
+    originalPresetsFile[presetType]!.push(newPreset);
+    await this.updatePresetsFile(originalPresetsFile);
+  }
+
+  private getIndentationSettings() {
+    const config = vscode.workspace.getConfiguration('editor', this.folder.uri);
+    const tabSize = config.get<number>('tabSize') || 4;
+    const insertSpaces = config.get<boolean>('insertSpaces') || true;
+    return { insertSpaces, tabSize };
+  }
+
+  async updatePresetsFile(presetsFile: preset.PresetsFile, isUserPresets = false): Promise<boolean> {
+    const presetsFilePath = isUserPresets? this.userPresetsPath : this.presetsPath;
+    const indent = this.getIndentationSettings();
+    try {
+        await fs.writeFile(presetsFilePath, JSON.stringify(presetsFile, null, indent.insertSpaces ? indent.tabSize : '\t'));
```

Comment:
When I created my first preset, I was expecting the CMakePresets.json file to open in the editor so I could confirm the values in the preset.  Should we open (or refocus) the file at this point?

Reply:
Sure we can do this